src/data/sym/card.lean

```diff
@@ -36,9 +35,336 @@ Prove the general case of stars and bars.
 stars and bars
 -/
 
+def multichoose1 (n k : ℕ) := fintype.card (sym (fin n) k)
+
+def multichoose2 (n k : ℕ) := (n + k - 1).choose k
+
+def encode (n k : ℕ) (x : sym (fin n.succ) k.succ) : sum (sym (fin n) k.succ) (sym (fin n.succ) k) :=
+  if h : ∃ y : fin n.succ, y ∈ x.val ∧ y = ⟨n, lt_add_one n⟩ then sum.inr ⟨x.val.erase ⟨n, lt_add_one n⟩, begin
+    cases h with a b,
+    cases b with c d,
+    have := multiset.card_erase_of_mem c,
+    rw d at this,
+    rw this,
+    rw x.property,
+    refl,
+  end⟩ else sum.inl ⟨x.val.map (λ a, ⟨if a.val = n then 0 else a.val, begin
+    have not_zero : n ≠ 0 := begin
+      intro g,
+      push_neg at h,
+      simp_rw g at h,
+      have := h 0 begin
+        cases n,
+        {
+          have is_zero : ∀ x : fin 1, x = 0 := begin
+            intro x,
+            cases x,
+            cases x_val,
+            refl,
+            norm_num at x_property,
+          end,
+          have zero_is_mem := @multiset.exists_mem_of_ne_zero _ x.val begin
+            have := x.property,
+            intro h,
+            rw h at this,
+            norm_num at this,
+          end,
+          cases zero_is_mem with w r,
+          rw (is_zero w) at r,
+          exact r,
+        },
+        {
+          norm_num at g,
+        },
+      end,
+      norm_num at this,
+    end,
+    split_ifs,
+    { exact pos_iff_ne_zero.mpr not_zero, },
+    {
+      have two_branches := lt_or_eq_of_le (nat.le_of_lt_succ a.property),
+      cases two_branches,
+      { assumption, },
+      {
+        exfalso,
+        exact h_1 two_branches,
+      },
+    },
+  end⟩), begin
+    rw multiset.card_map,
+    exact x.property,
+  end⟩
+
+def decode (n k : ℕ) : sum (sym (fin n) k.succ) (sym (fin n.succ) k) → sym (fin n.succ) k.succ := begin
+  intro x,
```

Comment:
Better to put this before the colon:
```suggestion
def decode (n k : ℕ) (x : sym (fin n) k.succ ⊕ sym (fin n.succ) k) : sym (fin n.succ) k.succ := begin
```
also better to use `⊕` instead of `sum` (this applies throughout

Reply:
Actually, I retract that...